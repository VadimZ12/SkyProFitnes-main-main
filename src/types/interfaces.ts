export interface User {
  uid: string;
  email: string;
  displayName?: string;
  photoURL?: string;
  name?: string;
  login?: string;
  password?: string;
  url_img?: string;
  customDisplayName?: string;
}

export interface Course {
  _id: string;
  nameEN: string;
  nameRU: string;
  description: string;
  directions: string[];
  fitting: string[];
  order: number;
  workouts: string[];
  imageUrl?: string;
  difficult?: number;
  totalDuration?: number;
  progress?: number;
}

export interface Workout {
  _id: string;
  name: string;
  video: string;
  exercises?: Exercise[];
}

export interface Exercise {
  name: string;
  quantity: number;
}

export type CachedData<T> = {
  data: T;
  timestamp: number;
};

interface ImportMetaEnv {
  VITE_FIREBASE_API_KEY: string;
  VITE_FIREBASE_AUTH_DOMAIN: string;
  VITE_FIREBASE_PROJECT_ID: string;
  VITE_FIREBASE_STORAGE_BUCKET: string;
  VITE_FIREBASE_MESSAGING_SENDER_ID: string;
  VITE_FIREBASE_APP_ID: string;
  VITE_FIREBASE_DATABASE_URL: string;
}

export interface ImportMeta {
  readonly env: ImportMetaEnv;
}