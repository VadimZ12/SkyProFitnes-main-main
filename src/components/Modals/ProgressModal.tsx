import React, { useState, useEffect } from 'react';
import { Exercise } from '../../types/interfaces';

interface ProgressModalProps {
  isOpen: boolean;
  onClose: () => void;
  exercises: Exercise[];
  onSave: (progress: { [key: string]: number }) => void;
  initialProgress: { [key: string]: number };
}

const ProgressModal: React.FC<ProgressModalProps> = ({ isOpen, onClose, exercises, onSave, initialProgress }) => {
  const [progress, setProgress] = useState<{ [key: string]: number }>(initialProgress);

  useEffect(() => {
    setProgress(initialProgress);
  }, [initialProgress]);

  if (!isOpen) return null;

  const handleInputChange = (exerciseName: string, value: string, maxQuantity: number) => {
    const numValue = Math.max(0, Math.min(parseInt(value) || 0, maxQuantity));
    setProgress(prev => ({ ...prev, [exerciseName]: numValue }));
  };

  const handleSubmit = (e: React.FormEvent) => {
    e.preventDefault();
    onSave(progress);
    onClose();
  };

  return (
    <div className="fixed inset-0 bg-black bg-opacity-50 flex items-center justify-center z-50">
      <div className="bg-white shadow-lg rounded-[30px] p-10 w-full max-w-[426px]">
        <h2 className="text-[32px] font-normal mb-[48px] text-center">Мой прогресс</h2>
        <form onSubmit={handleSubmit} className="flex flex-col gap-[20px]">
          {exercises.length > 0 ? (
            exercises.map((exercise) => (
              <div key={exercise.name} className="flex flex-col gap-[10px]">
                <label className="text-[18px]">{`Сколько раз вы сделали ${exercise.name}?`}</label>
                <input
                  type="text"
                  inputMode="numeric"
                  pattern="[0-9]*"
                  className="border border-[#D0CECE] rounded-[8px] p-[16px]"
                  placeholder="0"
                  value={progress[exercise.name] || ''}
                  onChange={(e) => handleInputChange(exercise.name, e.target.value, exercise.quantity)}
                />
              </div>
            ))
          ) : (
            <div className="flex flex-col gap-[10px]">
              <label className="text-[18px]">Завершили ли вы тренировку?</label>
              <select
                className="border border-[#D0CECE] rounded-[8px] p-[16px]"
                value={progress.completed ? "100" : "0"}
                onChange={(e) => setProgress({ completed: parseInt(e.target.value) })}
              >
                <option value="0">Нет</option>
                <option value="100">Да</option>
              </select>
            </div>
          )}
          <button type="submit" className="mt-[34px] bg-[#BCEC30] text-black rounded-[46px] py-[16px] px-[26px]">
            Сохранить
          </button>
        </form>
      </div>
    </div>
  );
};

export default ProgressModal;