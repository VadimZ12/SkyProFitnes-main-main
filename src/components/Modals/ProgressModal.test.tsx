import React from 'react'
import { render, fireEvent } from '@testing-library/react'
import ProgressModal from './ProgressModal'
import { Exercise } from '../../types/interfaces'

const mockExercises: Exercise[] = [
	{ name: 'Push Ups', quantity: 50 },
	{ name: 'Squats', quantity: 30 },
]

const mockInitialProgress = {
	'Push Ups': 20,
	Squats: 10,
}

describe('ProgressModal Snapshot', () => {
	it('renders the modal correctly when open', () => {
		const { asFragment } = render(
			<ProgressModal
				isOpen={true}
				onClose={() => {}}
				exercises={mockExercises}
				onSave={() => {}}
				initialProgress={mockInitialProgress}
			/>
		)
		expect(asFragment()).toMatchSnapshot()
	})

	it('renders nothing when the modal is closed', () => {
		const { asFragment } = render(
			<React.Fragment>
			<ProgressModal
				isOpen={false}
				onClose={() => {}}
				exercises={mockExercises}
				onSave={() => {}}
				initialProgress={mockInitialProgress}
			/>
			</React.Fragment>
		)
		expect(asFragment()).toMatchSnapshot()
	})

	it('calls onSave and onClose when form is submitted', () => {
		const mockOnSave = jest.fn()
		const mockOnClose = jest.fn()

		const { getByText } = render(
			<React.Fragment>
			<ProgressModal
				isOpen={true}
				onClose={mockOnClose}
				exercises={mockExercises}
				onSave={mockOnSave}
				initialProgress={mockInitialProgress}
			/>
			</React.Fragment>
		)

		const saveButton = getByText('Сохранить')
		fireEvent.click(saveButton)

		expect(mockOnSave).toHaveBeenCalledWith(mockInitialProgress)
		expect(mockOnClose).toHaveBeenCalled()
	})
})
