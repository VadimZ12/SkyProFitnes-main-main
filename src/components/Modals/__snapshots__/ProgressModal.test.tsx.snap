// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProgressModal Snapshot renders nothing when the modal is closed 1`] = `<DocumentFragment />`;

exports[`ProgressModal Snapshot renders the modal correctly when open 1`] = `
<DocumentFragment>
  <div
    class="fixed inset-0 bg-black bg-opacity-50 flex items-center justify-center z-50"
  >
    <div
      class="bg-white shadow-lg rounded-[30px] p-10 w-full max-w-[426px]"
    >
      <h2
        class="text-[32px] font-normal mb-[48px] text-center"
      >
        Мой прогресс
      </h2>
      <form
        class="flex flex-col gap-[20px]"
      >
        <div
          class="flex flex-col gap-[10px]"
        >
          <label
            class="text-[18px]"
          >
            Сколько раз вы сделали Push Ups?
          </label>
          <input
            class="border border-[#D0CECE] rounded-[8px] p-[16px]"
            inputmode="numeric"
            pattern="[0-9]*"
            placeholder="0"
            type="text"
            value="20"
          />
        </div>
        <div
          class="flex flex-col gap-[10px]"
        >
          <label
            class="text-[18px]"
          >
            Сколько раз вы сделали Squats?
          </label>
          <input
            class="border border-[#D0CECE] rounded-[8px] p-[16px]"
            inputmode="numeric"
            pattern="[0-9]*"
            placeholder="0"
            type="text"
            value="10"
          />
        </div>
        <button
          class="mt-[34px] bg-[#BCEC30] text-black rounded-[46px] py-[16px] px-[26px]"
          type="submit"
        >
          Сохранить
        </button>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
